src/lender/coordinator.sol

```diff
@@ -206,15 +214,49 @@ contract EpochCoordinator is Ticker, Auth  {
         executeEpoch(bestSubmission.seniorRedeem ,bestSubmission.juniorRedeem, bestSubmission.seniorSupply, bestSubmission.juniorSupply);
     }
 
+    function calcSeniorState(uint seniorRedeem, uint seniorSupply,uint seniorDebt, uint seniorBalance) public view returns (uint seniorAsset) {
+        return safeSub(safeAdd(safeAdd(seniorDebt, seniorBalance), seniorSupply), seniorRedeem);
+    }
+
+    function calcAssets(uint NAV, uint reserve_) public view returns(uint) {
+        return safeAdd(NAV, reserve_);
+    }
+
+
+    function calcSeniorRatio(uint seniorAsset, uint NAV, uint reserve_) public view returns(uint) {
+        uint assets = calcAssets(NAV, reserve_);
+        if(assets == 0) {
+            return 0;
+        }
+        return rdiv(seniorAsset, assets);
+    }
+
+    function reBalanceSeniorDebt(uint seniorAsset, uint currSeniorRatio) public view returns (uint seniorDebt_, uint seniorBalance_) {
+        uint seniorDebt = rmul(seniorAsset, currSeniorRatio);
+        return (seniorDebt, safeSub(seniorAsset, seniorDebt));
+    }
+
+    function calcFulfillment(uint amount, uint totalOrder) public view returns(uint percent) {
+        if(amount == 0 || totalOrder == 0) {
+            return 0;
+        }
+        return rdiv(amount, totalOrder);
+    }
+
     function executeEpoch(uint seniorRedeem, uint juniorRedeem, uint seniorSupply, uint juniorSupply) internal {
+        uint epochID = lastEpochExecuted+1;
```

Comment:
shouldn't we check whether submission period is over: `require(now > minChallengePeriodEnd)`
also it just makes sense to execute if `lastEpochExecuted < closingEpoch`

Reply:
the check is happening the the `public` executeEpoch method.